ext/LinearSolveHYPRE.jl

```diff
@@ -0,0 +1,193 @@
+const NO_COMM_SOLVERS = Union{HYPRE.BoomerAMG, HYPRE.Hybrid, HYPRE.ILU}
+create_solver(::Type{S}, comm) where {S <: NO_COMM_SOLVERS} = S()
+
+function create_solver(alg::HYPREAlgorithm, cache::LinearCache)
+    # If the solver is already instantiated, return it directly
+    if alg.solver isa HYPRE.HYPRESolver
+        return alg.solver
+    end
+
+    # Otherwise instantiate
+    if !(alg.solver <: Union{COMM_SOLVERS, NO_COMM_SOLVERS})
+        throw(ArgumentError("unknown or unsupported HYPRE solver: $(alg.solver)"))
+    end
+    comm = cache.cacheval.A.comm # communicator from the matrix
+    solver = create_solver(alg.solver, comm)
+
+    # Construct solver options
+    solver_options = (;
+                      AbsoluteTol = cache.abstol,
+                      MaxIter = cache.maxiters,
+                      PrintLevel = Int(cache.verbose),
+                      Tol = cache.reltol)
+
+    # Preconditioner (uses Pl even though it might not be a "left" preconditioner)
+    if !(cache.Pl isa Identity)
+        P = if cache.Pl isa HYPRESolver
+            cache.Pl
+        elseif cache.Pl <: HYPRESolver
+            create_solver(cache.Pl, comm)
+        else
+            throw(ArgumentError("unknown HYPRE preconditioner $(cache.Pl)"))
+        end
+        solver_options = merge(solver_options, (; Precond = P))
+    end
+
+    # Filter out some options that are not supported for some solvers
+    if solver isa HYPRE.Hybrid
+        # Rename MaxIter to PCGMaxIter
+        MaxIter = solver_options.MaxIter
+        ks = filter(x -> x !== :MaxIter, keys(solver_options))
+        solver_options = NamedTuple{ks}(solver_options)
+        solver_options = merge(solver_options, (; PCGMaxIter = MaxIter))
+    elseif solver isa HYPRE.BoomerAMG || solver isa HYPRE.ILU
+        # Remove AbsoluteTol, Precond
+        ks = filter(x -> !in(x, (:AbsoluteTol, :Precond)), keys(solver_options))
+        solver_options = NamedTuple{ks}(solver_options)
+    end
+
+    # Set the options
+    HYPRE.Internals.set_options(solver, pairs(solver_options))
+
+    return solver
+end
+
+# TODO: How are args... and kwargs... supposed to be used here?
+function SciMLBase.solve(cache::LinearCache, alg::HYPREAlgorithm, args...; kwargs...)
+    # It is possible to reach here without HYPRE.Init() being called if HYPRE structures are
+    # only to be created here internally (i.e. when cache.A::SparseMatrixCSC and not a
+    # ::HYPREMatrix created externally by the user). Be nice to the user and call it :)
+    if !(cache.A isa HYPREMatrix || cache.b isa HYPREVector || cache.u isa HYPREVector ||
+         alg.solver isa HYPRESolver)
+        HYPRE.Init()
+    end
+
+    # Move matrix and vectors to HYPRE, if not already provided as HYPREArrays
+    hcache = cache.cacheval
+    if hcache.A === nothing || cache.isfresh
+        hcache.A = cache.A isa HYPREMatrix ? cache.A : HYPREMatrix(cache.A)
+    end
+    isfresh_b = hcache.oidb != objectid(cache.b) # isfresh for b
```

Comment:
This isn't so nice, but I don't see another good way to know whether `b` has been changed. Perhaps overload `set_b(::LinearCache{..., HYPRECache, ...}, b)` would be one way?

I also think the same has to be done for `u`, since there is `set_u`.